### ISSUE TITLE
Mutable dataclass field checks don't respect immutable annotations

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#4038]
_Originally posted by @andersk in https://github.com/charliermarsh/ruff/issues/3877#issuecomment-1515536738_

**Comment by dhruvmanila** [comment]
I can take this up :)

Am I correct to assume that the said helper function in the original comment should be refactored to `ruff_python_semantic`? This is because it's dependent on `Context`.

**Comment by Pentusha** [comment]
I think my case is suitable for this issue:

error message:
```
 edgeql_qb/render/types.py:10:27: RUF009 Do not perform function call `FrozenDict` in dataclass defaults
```
 
immutable dataclass with immutable field:
```python
@dataclass(slots=True, frozen=True)
class RenderedQuery:
    query: str = ''
    context: FrozenDict = FrozenDict()
```

Hashable (Immutable) FrozenDict code [here](https://github.com/Pentusha/edgeql-qb/blob/master/edgeql_qb/frozendict.py#L7-L38).
For me it seems like false positive.

Ref: Pentusha/edgeql-qb#92

**Comment by dhruvmanila** [comment]
@Pentusha This comes from `RUF009` while this issue is to track `RUF008`. Yours is a valid case as well but it might be difficult to detect due to the lack of type inference.
